plan/plan.go

```diff
@@ -52,36 +52,47 @@ type Changes struct {
 // processing. It returns a copy of Plan with the changes populated.
 func (p *Plan) Calculate() *Plan {
 	changes := &Changes{}
-
-	// Ensure all desired records exist. For each desired record make sure it's
-	// either created or updated.
-	for _, desired := range p.Desired {
-		// Get the matching current record if it exists.
-		current, exists := recordExists(desired, p.Current)
-
-		// If there's no current record create desired record.
-		if !exists {
-			changes.Create = append(changes.Create, desired)
-			continue
+	usedDesired := map[int]bool{}
+	usedCurrent := map[int]bool{}
+
+	for c, current := range p.Current {
+		for d, desired := range p.Desired {
+			if current.DNSName == desired.DNSName && current.Target == desired.Target &&
+				current.SuitableType() == desired.SuitableType() {
+				usedDesired[d] = true
+				usedCurrent[c] = true
+				log.Debugf("Skipping endpoint %v because target has not changed", desired)
+			}
 		}
+	}
 
-		// If there already is a record update it if it changed.
-		if desired.Target != current.Target {
-			changes.UpdateOld = append(changes.UpdateOld, current)
-
-			desired.RecordType = current.RecordType // inherit the type from the dns provider
-			desired.MergeLabels(current.Labels)     // inherit the labels from the dns provider, including Owner ID
-			changes.UpdateNew = append(changes.UpdateNew, desired)
+	for c, current := range p.Current {
+		if _, found := usedCurrent[c]; found {
 			continue
 		}
-
-		log.Debugf("Skipping endpoint %v because target has not changed", desired)
+		for d, desired := range p.Desired {
```

Comment:
the point is, assume I have two ingresses, "foo.bar" -> "domain1", " foo.bar" -> "domain2". "foo.bar->domain1" is already registered on DNS server. Both will be treated as CNAME, both will belong to the same DNS zone, both will be passed to DNS provider (with this change) and therefore the whole batch will fail (including any other record), which make External-DNS useless until one of the ingresses is deleted. Current behaviour is to only create one of them, without failing the whole batch. Both behaviours are not desired and the question is which one we can tolerate for now